README.md
# Giganto: Raw-Event Storage System for AICE

Giganto is a high-performance raw-event storage system, specifically designed
for AICE. It is optimized to receive and store raw events through QUIC channels
and provides a flexible GraphQL API for querying the stored events. Giganto
empowers AICE with the ability to efficiently handle large-scale data
processing and real-time analytics.

[![Coverage Status](https://codecov.io/gh/aicers/giganto/branch/main/graph/badge.svg?token=AUUUIKX8O1)](https://codecov.io/gh/aicers/giganto)

## Features

* Scalable Storage: Giganto provides a scalable and distributed storage system,
  optimized for handling raw events generated by AICE sensors.
* GraphQL API: Giganto offers a powerful and flexible GraphQL API, enabling
  developers to query stored events with ease.
* QUIC Channels: Giganto utilizes the QUIC protocol to enable fast, low-latency
  communication and data transfer.
* High Performance: Giganto is designed to efficiently handle high volumes of
  data, ensuring optimal performance for AICE.

## Usage

You can run giganto by invoking the following command:

```sh
giganto <path to config file>
```

In the config file, you can specify the following options:

```toml
key = "key.pem"                            # path to private key file
cert = "cert.pem"                          # path to certificate file
roots = ["ca1.pem", "ca2.pem", "ca3.pem"]  # paths to CA certificate files
ingest_address = "0.0.0.0:38370"           # address to listen for ingest QUIC
publish_address = "0.0.0.0:38371"          # address to listen for publish QUIC
graphql_address = "127.0.0.1:8443"         # giganto's graphql address
data_dir = "tests/data"                    # path to directory to store data
retention = "100d"                         # retention period for data
log_dir = "/data/logs/apps"                # path to giganto's syslog file
export_dir = "tests/export"                # path to giganto's export file
max_open_files = 8000                      # db options max open files,
max_mb_of_level_base = 512                 # db options max MB of rocksDB Level 1
statistics_period = "10m"                  # statistics generate period
```

By default, giganto reads the config file from the following directories:

* Linux: `$HOME/.config/giganto/config.toml`
* macOS: `$HOME/Library/Application Support/com.einsis.giganto/config.toml`

For the `max_mb_of_level_base`, the last level has 100,000 times capacity,
and it is about 90% of total capacity. Therefore, about `db_total_mb / 111111` is
appropriate.
For example, `90`MB or less for 10TB Database, `900`MB or less for 100TB would
be appropriate.

These values assume you've used all the way up to level 6, so the actual values may
change if you want to grow your data further at the level base.
So if it's less than `512`MB, it's recommended to set default value of `512`MB.

## Test

Run giganto with the prepared configuration file. (Settings to use the
certificate/key from the tests folder.)

```sh
cargo run -- tests/config.toml
```

## License

Copyright 2022 EINSIS, Inc.

Licensed under [Apache License, Version 2.0][apache-license] (the "License");
you may not use this crate except in compliance with the License.

Unless required by applicable law or agreed to in writing, software distributed
under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See [LICENSE](LICENSE) for
the specific language governing permissions and limitations under the License.

## Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the [Apache-2.0
license][apache-license], shall be licensed as above, without any additional
terms or conditions.

[apache-license]: http://www.apache.org/licenses/LICENSE-2.0


CHANGELOG.md
# Changelog

This file documents recent notable changes to this project. The format of this
file is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and
this project adheres to [Semantic
Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Add `analyzer statistics` column family. generate and save traffic's
  raw data statistics.
- Add `request collector statistics`and `request analyzer statistics`
  to Graphql API.
- Delete DB by protocol name(column family) to graphQL API
  (without protocol will remove all data)

### Changed

- Modify `statistics` column family name to `collector statistics`

## [0.9.0] - 2023-04-03

### Added

- Add GraphQL API to return source list.

### Changed

- Change the format of data sent to REconverge from the publish module. ([ref](https://github.com/aicers/giganto-client/issues/9))

## [0.8.0] - 2023-03-30

### Added

- Add Giganto Restart processing.
- Add more fields to `dns`, `conn`, `http`
- Add common fields to network events.
- Publish support Packet request/response through QUIC.
- Add Packet store.
- Add GraphQL API for Packet.
- Add database options to config file.
- Add GraphQL API for config file.

### Changed

- Change field name `duration` to `last_time`.  (Except Conn struct)
- Modify to receive and process `multiple sources` of stream request messages
  from `HOG`.
- Modify module name `ingestion` to `ingest`.
- Create giganto's communication part as a separate crate. (giganto-client)
- Move init tracing to giganto-client crate for oplog logging.
- Fix packet logic in ingest.
- Rocksdb compression type has changed to Lz4, zstd from snappy
- Move giganto-client to separate repo [giganto-client](https://github.com/aicers/giganto-client).

## [0.7.0] - 2023-01-04

### Added

- Add export file to GraphQL API. (`csv`, `json` format support)
- Add `statistics` column family. Receive and save traffic statistics from Piglet.
- Save Giganto's `syslog` to a path written to `log_dir` in configuration file.
- Add `Oplog` (Operation log)

### Changed

- Send different stream start message depending on the daemon.
- Check the write permission on `/data/logs/apps` directory.

## [0.6.0] - 2022-12-06

### Added

- Publish support protocol record data request/response through QUIC.
- Add periodic time series to GraphQL API.
- Add send `all` source network stream to hog.
- Add more network data types. (`Ntlm`, `Kerberos`, `Ssh`, `DceRpc`)

### Changed

- The key in timeseries data includes timestamp.
- Change DNS answer field to `Vec<String>`.

### Removed

- Remove send network stream to hog from database.

### Fixed

- The subject name, rather than the issuer name, in the client certificate is
  used as an identifier.

## [0.5.0] - 2022-11-17

### Added

- Adds `smtp` to receive and save SMTP event.
- Adds SMTP to GraphQL API.

### Changed

- Adds `answer` field of DNS event.

## [0.4.0] - 2022-11-01

### Added

- Publish support network event(conn, dns, rdp, http) stream data through QUIC.
- GraphQL API returns all network raw events.

### Changed

- Separate compatibility version check criteria for publish and ingestion.

## [0.3.0] - 2022-10-20

### Added

- Support periodic time series.
- GraphQL API supports filtering with source name, kind name of log, time range,
  IP address range, port range.
- Publish check protocol compatibility with connected Programs.
- Publish support log/period time series record data request/response through QUIC.
- Add Packets Request GraphQL API.
- Send acknowledgment of channel done messages sent by reproduce.

### Changed

- GraphQL API now accepts `filter`, which includes the source name, time range,
  IP address ranges, and port ranges.

### Fixed

- Fixed a bug returning wrong events when the time range is specified.

## [0.2.0] - 2022-10-04

### Added

- limit Maximum request page size. (Requests over 100 will be treated as 100)
- Check protocol compatibility with connected Programs.
- Accepts QUIC connections from subscribers.
- GraphQL API supports cursor-based pagination.
- GraphQL API takes `start` and `end` parameters to specify the time range of
  the query.

### Fixed

- A query for a certain source no longer returns events from other sources.

## [0.1.0] - 2022-09-16

### Added

- Initial release.

[0.9.0]: https://github.com/aicers/giganto/compare/0.8.0...0.9.0
[0.8.0]: https://github.com/aicers/giganto/compare/0.7.0...0.8.0
[0.7.0]: https://github.com/aicers/giganto/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/aicers/giganto/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/aicers/giganto/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/aicers/giganto/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/aicers/giganto/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/aicers/giganto/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/aicers/giganto/tree/0.1.0


src/settings.rs
//! Configurations for the application.
use config::{builder::DefaultState, Config, ConfigBuilder, ConfigError, File};
use serde::{de::Error, Deserialize, Deserializer};
use std::{net::SocketAddr, path::PathBuf, time::Duration};

const DEFAULT_INGEST_ADDRESS: &str = "[::]:38370";
const DEFAULT_PUBLISH_ADDRESS: &str = "[::]:38371";
const DEFAULT_GRAPHQL_ADDRESS: &str = "[::]:8443";

/// The application settings.
#[derive(Clone, Debug, Deserialize)]
pub struct Settings {
    pub cert: PathBuf,       // Path to the certificate file
    pub key: PathBuf,        // Path to the private key file
    pub roots: Vec<PathBuf>, // Path to the rootCA file
    #[serde(deserialize_with = "deserialize_socket_addr")]
    pub ingest_address: SocketAddr, // IP address & port to ingest data
    #[serde(deserialize_with = "deserialize_socket_addr")]
    pub publish_address: SocketAddr, // IP address & port to publish data
    pub data_dir: PathBuf,   // DB storage path
    #[serde(with = "humantime_serde")]
    pub retention: Duration, // Data retention period
    #[serde(deserialize_with = "deserialize_socket_addr")]
    pub graphql_address: SocketAddr, // IP address & port to graphql
    pub log_dir: PathBuf,    //giganto's syslog path
    pub export_dir: PathBuf, //giganto's export file path
    // db options
    pub max_open_files: i32,
    pub max_mb_of_level_base: u64,
    #[serde(with = "humantime_serde")]
    pub statistics_period: Duration, // statistics generate period
}

impl Settings {
    /// Creates a new `Settings` instance, populated from the default
    /// configuration file if it exists.
    pub fn new() -> Result<Self, ConfigError> {
        let dirs = directories::ProjectDirs::from("com", "einsis", "giganto").expect("unreachable");
        let config_path = dirs.config_dir().join("config.toml");
        if config_path.exists() {
            // `config::File` requires a `&str` path, so we can't use `config_path` directly.
            if let Some(path) = config_path.to_str() {
                Self::from_file(path)
            } else {
                Err(ConfigError::Message(
                    "config path must be a valid UTF-8 string".to_string(),
                ))
            }
        } else {
            default_config_builder().build()?.try_deserialize()
        }
    }

    /// Creates a new `Settings` instance, populated from the given
    /// configuration file.
    pub fn from_file(cfg_path: &str) -> Result<Self, ConfigError> {
        let s = default_config_builder()
            .add_source(File::with_name(cfg_path))
            .build()?;

        s.try_deserialize()
    }
}

/// Creates a new `ConfigBuilder` instance with the default configuration.
fn default_config_builder() -> ConfigBuilder<DefaultState> {
    let dirs = directories::ProjectDirs::from("com", "einsis", "giganto").expect("unreachable");
    let db_dir =
        directories::ProjectDirs::from_path(PathBuf::from("db")).expect("unreachable db dir");
    let log_dir = directories::ProjectDirs::from_path(PathBuf::from("logs/apps"))
        .expect("unreachable logs dir");
    let export_dir = directories::ProjectDirs::from_path(PathBuf::from("export"))
        .expect("unreachable export dir");
    let db_path = db_dir.data_dir().to_str().expect("unreachable db path");
    let log_path = log_dir.data_dir().to_str().expect("unreachable log path");
    let export_path = export_dir
        .data_dir()
        .to_str()
        .expect("unreachable export path");
    let config_dir = dirs.config_dir();
    let cert_path = config_dir.join("cert.pem");
    let key_path = config_dir.join("key.pem");

    Config::builder()
        .set_default("cert", cert_path.to_str().expect("path to string"))
        .expect("default cert dir")
        .set_default("key", key_path.to_str().expect("path to string"))
        .expect("default key dir")
        .set_default("ingest_address", DEFAULT_INGEST_ADDRESS)
        .expect("valid address")
        .set_default("publish_address", DEFAULT_PUBLISH_ADDRESS)
        .expect("valid address")
        .set_default("graphql_address", DEFAULT_GRAPHQL_ADDRESS)
        .expect("local address")
        .set_default("data_dir", db_path)
        .expect("data dir")
        .set_default("retention", "100d")
        .expect("retention")
        .set_default("log_path", log_path)
        .expect("log dir")
        .set_default("export_path", export_path)
        .expect("export_dir")
        .set_default("max_open_files", 8000)
        .expect("default max open files")
        .set_default("max_mb_of_level_base", 512)
        .expect("default max mb of level base")
        .set_default("statistics_period", "1m")
        .expect("statistics period")
}

/// Deserializes a socket address.
///
/// # Errors
///
/// Returns an error if the address is not in the form of 'IP:PORT'.
fn deserialize_socket_addr<'de, D>(deserializer: D) -> Result<SocketAddr, D::Error>
where
    D: Deserializer<'de>,
{
    let addr = String::deserialize(deserializer)?;
    addr.parse()
        .map_err(|e| D::Error::custom(format!("invalid address \"{addr}\": {e}")))
}
